### PR TITLE
Restore per-context layer-output shape via configurable selective k-CFA

### PR DESCRIPTION
**Pull request by khatchad**
## Summary

A user `tf.keras.Model` subclass invoked from multiple sites (e.g. a training loop and a test-set evaluation) merged into one context-insensitive call-graph node, collapsing its per-caller layer-output allocations and losing per-context tensor shape. Downstream, `tf.argmax` of a collapsed layer output leaked one caller's shape into another context and threw a `NonBroadcastableShapesException` in `tf.equal`, which is why `Argmax.getDefaultShapes` returned ⊤ to sidestep the regression.

This PR resolves that by making the targeted k-CFA depth configurable (https://github.com/wala/ML/issues/379) and applying deeper context to user model forward methods (https://github.com/wala/ML/issues/530), then restoring the precise `tf.argmax`/`tf.argmin` output shape.

## Changes

- **Configurable depth (https://github.com/wala/ML/issues/379):** the hard-coded `2` on the targeted context selector becomes `DEFAULT_TARGETED_CFA_DEPTH` (default 2, fully back-compat), exposed as a `PythonTensorAnalysisEngine` constructor parameter. Clients supply a deeper value when allocations merge across call sites; only the *targeted* selector (framework APIs plus model methods) uses it, never the whole program.
- **Model-aware context (https://github.com/wala/ML/issues/530):** the targeted predicate is extended to user `tf.keras.Model` forward methods so a model called from multiple sites is analyzed per caller, keeping its layer-output allocations distinct.
- **Precise `tf.argmax`/`tf.argmin` shape:** `Argmax.getDefaultShapes` now delegates to `ReduceMean`'s keepdims=false reduction (input shape with the `axis` dimension removed), since the per-context separation removes the broadcast regression that previously forced ⊤.
- **Test harness:** depth-parameterized `makeEngine`/`test` overloads (mirroring the analogous knob in the Java streams tooling), so `testNeuralNetwork1-4` opt into depth 4 while every other test stays at the default.

## Detection: Heuristic vs CHA-subtype (Dormant Path)

The clean way to recognize a user model method is a class-hierarchy subtype check against `tf.keras.Model`. That check is **wired but intentionally dormant**: the front-end does not yet record `class X(Model)` inheritance, so a user model class resolves its superclass to `Lobject` rather than the Keras `Model` type (filed as https://github.com/wala/ML/issues/571, an existing front-end TODO). Until that lands, a structural heuristic over user `call`/`__call__`/`predict` methods drives the predicate. When https://github.com/wala/ML/issues/571 is resolved, the heuristic is deleted and `isModelSubclassMethod` takes over—both are present so the switch is a one-line change.

## Testing

- Full suite green (all modules, `mvn test`).
- `testNeuralNetwork1-4` run at depth 4: `accuracy`'s `y_pred` is now the correct per-context union `{(256, 10) float32, ? float32}`, and the per-context multiplicity for `testNeuralNetwork` rises to 24.
- The six `tf.argmax`/`tf.argmin` shape tests tighten from ⊤ to the precise `(3,)` the fixtures assert at runtime.

Closes https://github.com/wala/ML/issues/379.
Closes https://github.com/wala/ML/issues/530.

### CODE REVIEW

File: com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/PythonTensorAnalysisEngine.java [hunk 1]
```diff
@@ -84,12 +98,38 @@ public PythonTensorAnalysisEngine(List<File> pythonPath) {
...
+   * @param targetFramework The framework name prefix whose API methods receive deep context.
+   * @param targetedCfaDepth The k-CFA depth to apply (e.g. {@link #DEFAULT_TARGETED_CFA_DEPTH});
+   *     higher values increase precision at the cost of analysis time.
+   */
+  public PythonTensorAnalysisEngine(
+      List<File> pythonPath, String targetFramework, int targetedCfaDepth) {
     super(pythonPath);
     this.targetFramework = targetFramework;
+    this.targetedCfaDepth = targetedCfaDepth;
   }
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/PythonTensorAnalysisEngine.java:151]: > `targetedCfaDepth` is accepted from callers and passed directly into `nCFAContextSelector` without validation. A negative depth is invalid for k-CFA and can fail at runtime in a way that's hard to diagnose; it's better to reject it early with an `IllegalArgumentException` (and ideally document the valid range).

Good catch. Added a shared `checkTargetedCfaDepth` guard that rejects a negative depth with an `IllegalArgumentException` from both depth-taking constructors, documented the non-negative range on the `@param`/`@throws`, and covered it with `testNegativeTargetedCfaDepthRejected`. Fixed in 38cf0ee6.

File: com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/Argmax.java [hunk 2]
```diff
@@ -85,7 +87,10 @@ public Argmax(PointsToSetVariable source) {
 
   @Override
   protected Set<List<Dimension<?>>> getDefaultShapes(PropagationCallGraphBuilder builder) {
-    return null;
+    // argmax removes the `axis` dimension (no keepdims), which is exactly ReduceMean's
+    // keepdims=False reduction. With per-context layer-output allocations (wala/ML#530), the input
+    // shape is resolved per caller, so the result no longer collapses across contexts.
+    return super.getDefaultShapes(builder);
   }
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/Argmax.java:122]: > `Argmax` now delegates shape inference to `ReduceMean`, but `ReduceMean` only recognizes the `axis` keyword. `tf.math.argmax` also supports the deprecated `dimension` alias (already modeled in `Parameters.DIMENSION`), and with the current implementation a call like `tf.argmax(x, dimension=0)` will be mis-modeled as `axis=None` (scalar) instead of removing the specified axis. Consider handling `dimension` explicitly and hard-coding keepdims=false semantics (since argmax has no keepdims parameter).

Correct. I attempted an interim guard that falls back to ⊤ when the alias is present, but `DIMENSION` isn't a positionally-modeled parameter in the `tensorflow.xml` summary, so the positional probe is unreliable (it tripped the common `axis=0` path to ⊤). Rather than work around the summary modeling here, I filed wala/ML#572 to honor the alias precisely with keyword-aware axis resolution, and captured the keepdims/`output_type` index-overlap robustness note there too. Documented the limitation in `Argmax.getDefaultShapes` in 38cf0ee6.

File: com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/Argmax.java [hunk 3]
```diff
@@ -85,7 +87,14 @@ public Argmax(PointsToSetVariable source) {
 
   @Override
   protected Set<List<Dimension<?>>> getDefaultShapes(PropagationCallGraphBuilder builder) {
-    return null;
+    // argmax removes the `axis` dimension (no keepdims), which is exactly ReduceMean's
+    // keepdims=False reduction. With per-context layer-output allocations (wala/ML#530), the input
+    // shape is resolved per caller, so the result no longer collapses across contexts.
+    //
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/Argmax.java:117]: > `Argmax` delegates shape inference to `ReduceMean#getDefaultShapes`, but `ReduceMean` interprets parameter position 2 as `keepdims`. For `tf.math.argmax`, position 2 is `output_type`, so positional calls like `tf.math.argmax(x, 0, tf.int32)` will be mis-modeled as `keepdims` and can incorrectly yield a union that includes keepdims=true shapes (e.g. `(1, 3)`), which is wrong for argmax/argmin (they never keep dims).

Good catch, and you're right that it's not merely theoretical: a dtype constant is neither boolean nor numeric, so `ReduceMean` widens `keepdims` to `{false, true}` and unions a spurious `keepdims=true` shape. Fixed in cb7c377a by extracting `ReduceMean.getKeepDimsValues` and overriding it in `Argmax` (and `Argmin`) to force `{false}`, since they have no `keepdims` parameter. Added `testArgmaxOutputTypePositional` (`tf.math.argmax(x, 0, tf.int32)`) as a regression guard, asserting the precise `(3,) int32` with no `(1, 3)`.

File: com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/Argmax.java [hunk 4]
```diff
@@ -85,7 +87,14 @@ public Argmax(PointsToSetVariable source) {
...
+    // ReduceMean resolves the reduction axis from the `axis` parameter only, so argmax's deprecated
+    // TF 1.x `dimension` alias is not honored for shape (a `tf.argmax(x, dimension=0)` call would
+    // reduce as `axis=None`). Honoring the alias precisely is tracked by wala/ML#572.
+    return super.getDefaultShapes(builder);
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/Argmax.java:121]: > `Argmax` delegates shape inference to `ReduceMean#getDefaultShapes`, but `ReduceMean` interprets parameter position 2 as `keepdims`. For `tf.math.argmax`, position 2 is `output_type`, so positional calls like `tf.math.argmax(x, 0, tf.int32)` will be mis-modeled as `keepdims` and can incorrectly yield a union that includes keepdims=true shapes (e.g. `(1, 3)`), which is wrong for argmax/argmin (they never keep dims).

Good catch, and you're right that it's not merely theoretical: a dtype constant is neither boolean nor numeric, so `ReduceMean` widens `keepdims` to `{false, true}` and unions a spurious `keepdims=true` shape. Fixed in cb7c377a by extracting `ReduceMean.getKeepDimsValues` and overriding it in `Argmax` (and `Argmin`) to force `{false}`, since they have no `keepdims` parameter. Added `testArgmaxOutputTypePositional` (`tf.math.argmax(x, 0, tf.int32)`) as a regression guard, asserting the precise `(3,) int32` with no `(1, 3)`.

File: com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/ReduceMean.java [hunk 5]
```diff
@@ -43,26 +43,22 @@ public ReduceMean(PointsToSetVariable source) {
...
+   * force {@code {false}}, which also prevents misreading an argument that sits at the same
+   * positional index as {@code keepdims} (see {@link Argmax}).
+   *
+   * @param builder The {@link PropagationCallGraphBuilder} used to build the call graph.
+   * @return The set of possible {@code keepdims} values; never empty.
+   */
+  protected Set<Boolean> getKeepDimsValues(PropagationCallGraphBuilder builder) {
     OrdinalSet<InstanceKey> keepDimsPts =
         this.getArgumentPointsToSet(
             builder, Parameters.KEEPDIMS.getIndex(), Parameters.KEEPDIMS.getName());
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/ReduceMean.java:81]: > `getKeepDimsValues` Javadoc says a non-boolean, non-numeric *constant* keepdims value widens to `{false, true}` for soundness, but the current `ConstantKey` branch only handles `Boolean`, `Number`, and `null` and otherwise silently ignores the constant, which can collapse to `{false}` only.

Fixed in b6b760a3 — added the missing `else` so a non-boolean, non-numeric constant widens to `{false, true}`, matching the Javadoc and the non-constant case.

File: com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/Argmax.java [hunk 6]
```diff
@@ -85,7 +87,30 @@ public Argmax(PointsToSetVariable source) {
 
   @Override
   protected Set<List<Dimension<?>>> getDefaultShapes(PropagationCallGraphBuilder builder) {
-    return null;
+    // argmax removes the `axis` dimension (no keepdims), which is exactly ReduceMean's
+    // keepdims=False reduction. With per-context layer-output allocations (wala/ML#530), the input
+    // shape is resolved per caller, so the result no longer collapses across contexts.
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/Argmax.java:116]: > `Argmax#getDefaultShapes` delegates to `ReduceMean#getDefaultShapes`, but that method resolves the input argument by the keyword name `input_tensor`. In `tensorflow.xml`, `argmax`'s first parameter is named `input`, so a call like `tf.math.argmax(input=x, axis=0)` (keyword-only `input`) will fail argument resolution and can throw `IllegalStateException` during analysis.

Confirmed and fixed in be43d952. Rather than recomputing the shape in `Argmax` (which would duplicate `ReduceMean`'s reduction), I made the input-parameter name and position overridable in `ReduceMean` and overrode them in `Argmax` to its `input` parameter, keeping the reduction in one place. `tf2_test_argmax_input_keyword.py` and `testArgmaxInputKeyword` guard the keyword-`input` case.
